File: object_analytics_visualization/object_analytics_visualization/marker_publisher.py
# Copyright (c) 2018 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from copy import deepcopy


from geometry_msgs.msg import Point
from geometry_msgs.msg import Pose

import message_filters

from object_analytics_msgs.msg import ObjectsInBoxes3D
from object_analytics_msgs.msg import TrackedObjects
from object_msgs.msg import ObjectsInBoxes

import rclpy

from rclpy.node import Node
from std_msgs.msg import ColorRGBA
from visualization_msgs.msg import Marker
from visualization_msgs.msg import MarkerArray


class Roi(object):
    """
    Wrapper of sensor_msgs.msg.RegionOfInterest for hashing.

    when roi is used to be as dict key, it's required to have its own hash implementation
    which needs to implement two special methods __hash__ and __eq__

    """

    def __init__(self, roi):
        """
        Construct a Roi from sensor_msgs.msg.RegionOfInterest instance.

        Args:
            roi (RegionOfInterest): Roi of a detected object or tracked object

        """
        self._hash = hash((roi.x_offset, roi.y_offset, roi.height, roi.width))

    def __hash__(self):
        return self._hash

    def __eq__(self, other):
        return hash(self) == hash(other)


class ObjectItem(object):
    """
    An ObjectItem represents a merged result.

    Merge localization, detection and tracking which has the same roi.
    The main usage of this class is to build Marker based on the merged information.

    """

    def __init__(self, header, roi, track_id, detected_object, min_3d, max_3d):
        """
        Build an instance from messages.

        Args:
            header (std_msgs.msg.Header):      Message header
            roi (RegionOfInterest):            Region of interest
            track_id (int):                    Tracking id
            detected_object (Object):          Instance of object_msgs.msg.Object
            min_3d (geometry_msgs.msg.Point32):   Min position in 3d space
            max (geometry_msgs.msg.Point32):   Max position in 3d space

        """
        self.YELLOW = ColorRGBA()
        self.YELLOW.r = 1.0
        self.YELLOW.g = 1.0
        self.YELLOW.b = 0.0
        self.YELLOW.a = 1.0

        self.GREEN = ColorRGBA()
        self.GREEN.r = 0.0
        self.GREEN.g = 1.0
        self.GREEN.b = 0.0
        self.GREEN.a = 1.0

        self.POSE = Pose()
        self._header = header
        self._roi = roi
        self._track_id = track_id
        self._object = detected_object

        self._p1 = Point()
        self._p1.x = min_3d.x
        self._p1.y = min_3d.y
        self._p1.z = min_3d.z

        self._p2 = deepcopy(self._p1)
        self._p2.x = max_3d.x

        self._p3 = deepcopy(self._p2)
        self._p3.z = max_3d.z

        self._p4 = deepcopy(self._p3)
        self._p4.x = min_3d.x

        self._p5 = deepcopy(self._p1)
        self._p5.y = max_3d.y

        self._p6 = deepcopy(self._p5)
        self._p6.x = max_3d.x

        self._p7 = deepcopy(self._p6)
        self._p7.z = max_3d.z

        self._p8 = deepcopy(self._p7)
        self._p8.x = min_3d.x

    def linelist(self):
        """Build line list marker from 8 points in 3d space."""
        line_list = Marker()
        line_list.header = self._header
        line_list.type = Marker.LINE_LIST
        line_list.action = Marker.ADD
        line_list.scale.x = 0.005
        line_list.color = self.YELLOW
        line_list.pose = deepcopy(self.POSE)

        line_list.pose.position.x = self._p1.x
        line_list.pose.position.y = self._p1.y
        line_list.pose.position.z = self._p1.z

        line_list.points.extend((self._p1, self._p2))
        line_list.points.extend((self._p2, self._p3))
        line_list.points.extend((self._p3, self._p4))
        line_list.points.extend((self._p4, self._p1))
        line_list.points.extend((self._p5, self._p6))
        line_list.points.extend((self._p6, self._p7))
        line_list.points.extend((self._p7, self._p8))
        line_list.points.extend((self._p8, self._p5))
        line_list.points.extend((self._p1, self._p5))
        line_list.points.extend((self._p2, self._p6))
        line_list.points.extend((self._p3, self._p7))
        line_list.points.extend((self._p4, self._p8))

        return line_list

    def min_text(self):
        """Build text marker to hold min_3d text."""
        text = Marker()
        text.header = self._header
        text.type = Marker.TEXT_VIEW_FACING
        text.action = Marker.ADD
        text.scale.z = 0.05
        text.color = self.GREEN
        text.pose = deepcopy(self.POSE)
        text.pose.position.x = self._p1.x
        text.pose.position.y = self._p1.y
        text.pose.position.z = self._p1.z
        text.text = 'Min[{:.2f} {:.2f} {:.2f}]'.format(self._p1.x, self._p1.y, self._p1.z)
        return text

    def max_text(self):
        """Build text marker to hold max_3d text."""
        text = Marker()
        text.header = self._header
        text.type = Marker.TEXT_VIEW_FACING
        text.action = Marker.ADD
        text.scale.z = 0.05
        text.color = self.GREEN
        text.pose = deepcopy(self.POSE)
        text.pose.position.x = self._p7.x
        text.pose.position.y = self._p7.y
        text.pose.position.z = self._p7.z
        text.text = 'Max[{:.2f} {:.2f} {:.2f}]'.format(self._p7.x, self._p7.y, self._p7.z)
        return text

    def name_id_text(self):
        """Build text marker to hold name and tracking id."""
        text = Marker()
        text.header = self._header
        text.type = Marker.TEXT_VIEW_FACING
        text.action = Marker.ADD
        text.scale.z = 0.05
        text.color = self.GREEN
        text.pose = deepcopy(self.POSE)
        text.pose.position.x = self._p1.x
        text.pose.position.y = (self._p1.y + self._p5.y) / 2
        text.pose.position.z = self._p1.z
        text.text = '{} #{}'.format(self._object.object_name, self._track_id)
        return text


class SynchronizedSubscriber(Node):
    """Time synchronizer wrapper for localization."""

    def __init__(self):
        """Time synchronizer for Localization, detection and tracking."""
        super().__init__('object_analytics_visualization3d')

        self.get_logger().info('Start marker_publisher')
        self._pub = self.create_publisher(MarkerArray, '/object_analytics/localization_rviz')

        loc_sub = message_filters.Subscriber(self, ObjectsInBoxes3D,
                                             '/object_analytics/localization')
        det_sub = message_filters.Subscriber(self, ObjectsInBoxes,
                                             '/movidius_ncs_stream/detected_objects')
        tra_sub = message_filters.Subscriber(self, TrackedObjects,
                                             '/object_analytics/tracking')

        ts = message_filters.ApproximateTimeSynchronizer([det_sub, loc_sub, tra_sub], 10, 0.01)

        ts.registerCallback(self._callback)

    def _callback(self, detected_objects, localized_objects, tracked_objects):
        """
        Time synchronizer callback.

        Args:
            detected_objects (object_msgs.msg.ObjectsInBoxes):
                              Detection results of current frame
            localized_object (object_analytics_msgs.msg.ObjectsInBoxes3D):
                              Localization results of current frame
            tracked_objects (object_analytics_msgs.msg.TrackedObjects):
                              Tracking results of current frame

        """
        self.get_logger().info('****************TimeSynchromizer: callback**************')
        objects = self._merge_results(detected_objects, localized_objects, tracked_objects)
        marker_array = self._build_marker_array(detected_objects.header, objects)
        self.get_logger().info('[publish]: "%s"' % marker_array)
        self._pub.publish(marker_array)

    # @staticmethod
    # def _merge_results(self, detected_objects, localized_objects):
    def _merge_results(self, detected_objects, localized_objects, tracked_objects):
        """
        Merge detection, localization and tracking which the same roi.

        roi should be include in localizaton result
        Args:
            detected_objects (object_msgs.msg.ObjectsInBoxes):
                              Detection results of current frame
            localized_object (object_analytics_msgs.msg.ObjectsInBoxes3D):
                              Localization results of current frame
            tracked_objects (object_analytics_msgs.msg.TrackedObjects):
                             Tracking results of current frame
        Returns:
            ObjectItem list: list of ObjectItem in which each item is merged from
                             detction, tracking and localization.

        """
        key = []
        localized_map = {}
        for obj in localized_objects.objects_in_boxes:
            localized_map[Roi(obj.roi)] = obj

        detected_map = {}
        for obj in detected_objects.objects_vector:
            key = Roi(obj.roi)
            if key not in localized_map:
                continue
            detected_map[Roi(obj.roi)] = obj

        if key is None:
            return

        merged = []
        for obj in tracked_objects.tracked_objects:
            # TODO: later to fix track issue and replace with roi check, not only x_offset check.
            # for obj in tracked_objects.tracked_objects:
                # key = Roi(obj.roi)
            if obj.roi.x_offset != localized_map[key].roi.x_offset:
                continue
            merged.append(ObjectItem(tracked_objects.header, obj.roi, obj.id,
                                     detected_map[key].object,
                                     localized_map[key].min, localized_map[key].max))
        return merged

    # @staticmethod
    def _build_marker_array(self, header, objects):
        """
        Build MarkerArray from given header and list of ObjectItem.

        Args:
            header (std_msgs.msg.Header):
                     Header of current processing frame
            objects (list):
                     List of ObjectItem in which each item represents a merged result
        Returns:
            MarkerArray: Includes marker for cleaning preivous markers,
                         marker for bouding box and marker for text.

        """
        marker_array = MarkerArray()

        clean_all_marker = Marker()
        clean_all_marker.header = header
        clean_all_marker.action = Marker.DELETEALL
        marker_array.markers.append(clean_all_marker)

        markers = []
        for obj in objects:
            markers.extend([obj.linelist()])
            markers.extend([obj.min_text()])
            markers.extend([obj.max_text()])
            markers.extend([obj.name_id_text()])
        for idx, marker in enumerate(markers):
            marker.id = idx
        marker_array.markers.extend(markers)

        return marker_array


def main(args=None):
    rclpy.init(args=args)

    synchronized_subscriber = SynchronizedSubscriber()

    rclpy.spin(synchronized_subscriber)

    synchronized_subscriber.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: object_analytics_rviz/src/marker_publisher.cpp
// Copyright (c) 2018 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <object_msgs/msg/objects_in_boxes.hpp>
#include <geometry_msgs/msg/point.hpp>
#include <visualization_msgs/msg/marker_array.hpp>

#include <object_analytics_msgs/msg/tracked_objects.hpp>
#include <object_analytics_msgs/msg/objects_in_boxes3_d.hpp>
#include <object_analytics_msgs/msg/tracked_object.hpp>

#include <chrono>
#include <string>
#include <vector>
#include <memory>

using namespace std::chrono_literals;
using std::placeholders::_1;

using TrackingMsg = object_analytics_msgs::msg::TrackedObjects;
using LocalizationMsg = object_analytics_msgs::msg::ObjectsInBoxes3D;
using TrackingObjectInBox = object_analytics_msgs::msg::TrackedObject;
using LocalizationObjectInBox = object_analytics_msgs::msg::ObjectInBox3D;

const int kMsgQueueSize = 10;
/* This demo code is desiged for showing object analytics result on rviz.
 * Subscribe localization/tracking msg, publish box_3d_markers for display. */

class MarkerPublisher : public rclcpp::Node
{
public:
  MarkerPublisher()
  : Node("marker_publisher")
  {
    loc_performance_subscription_ = this->create_subscription<LocalizationMsg>(
      "/object_analytics/localization",
      std::bind(&MarkerPublisher::loc_performance_callback, this, _1));
    loc_marker_subscription_ = this->create_subscription<LocalizationMsg>(
      "/object_analytics/localization", std::bind(&MarkerPublisher::loc_marker_callback, this, _1));
    marker_pub_ =
      create_publisher<visualization_msgs::msg::MarkerArray>("/object_analytics/marker_publisher");


    RCLCPP_INFO(get_logger(), "Start MarkerPublisher ...");
  }

private:
  using ObjectRoi = sensor_msgs::msg::RegionOfInterest;

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Subscription<LocalizationMsg>::SharedPtr loc_marker_subscription_;
  rclcpp::Subscription<LocalizationMsg>::SharedPtr loc_performance_subscription_;
  rclcpp::Subscription<TrackingMsg>::SharedPtr tra_subscription_;
  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_pub_;

  float loc_latency_;
  float loc_fps_;
  float tra_latency_;
  float tra_fps_;

  /* create 3d boxes of objects */
  void loc_marker_callback(
    const LocalizationMsg::SharedPtr loc)
  {
    if (loc->objects_in_boxes.size() != 0) {
      std::vector<LocalizationObjectInBox> objects_localized;
      std_msgs::msg::Header header = loc->header;
      objects_localized = loc->objects_in_boxes;
      MarkerPublisher::createMarker(header, objects_localized);
    }
  }

  void createMarker(
    std_msgs::msg::Header header,
    std::vector<LocalizationObjectInBox> loc_objects)
  {
    visualization_msgs::msg::MarkerArray marker_array_loc;
    visualization_msgs::msg::Marker marker_clear;
    marker_array_loc.markers = std::vector<visualization_msgs::msg::Marker>();
    marker_clear.action = visualization_msgs::msg::Marker::DELETEALL;
    marker_clear.header = header;
    marker_array_loc.markers.emplace_back(marker_clear);
    int marker_id = 0;
    for (auto loc : loc_objects) {
      if (loc.min.x == 0 && loc.min.y == 0 && loc.min.z == 0 &&
        loc.max.x == 0 && loc.max.y == 0 && loc.max.z == 0)
      {
        break;
      }
      geometry_msgs::msg::Point box_min;
      box_min.x = loc.min.x;
      box_min.y = loc.min.y;
      box_min.z = loc.min.z;
      geometry_msgs::msg::Point box_max;
      box_max.x = loc.max.x;
      box_max.y = loc.max.y;
      box_max.z = loc.max.z;
      std::string obj_name = loc.object.object_name;
      MarkerPublisher::addMarker(marker_array_loc, header, box_min, box_max,
        obj_name, marker_id);
    }
    MarkerPublisher::addPerformanceMarker(marker_array_loc, header, loc_fps_, loc_latency_,
      ++marker_id);
    marker_pub_->publish(marker_array_loc);
  }

  void addPerformanceMarker(
    visualization_msgs::msg::MarkerArray & marker_array,
    std_msgs::msg::Header header, float loc_fps_, float loc_latency_,
    int & marker_id)
  {
    auto marker = visualization_msgs::msg::Marker();
    marker.header = header;
    marker.id = marker_id;
    marker.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
    marker.action = visualization_msgs::msg::Marker::ADD;

    char performance_text[100];
    snprintf(performance_text, sizeof(performance_text), "Localization:fps=%.2fHz,latency=%.2fSec",
      loc_fps_, loc_latency_);
    marker.scale.z = 0.05;
    marker.color.a = 1.0;
    marker.color.r = 1.0;
    marker.color.g = 0.0;
    marker.color.b = 0.0;
    marker.text = performance_text;
    marker.pose.position.x = 0;
    marker.pose.position.y = -0.5;
    marker.pose.position.z = 0.6;
    marker_array.markers.emplace_back(marker);
  }
  /* add the marker composed by object_name, object_id, mix points, max points, 3d box bounaries*/
  void addMarker(
    visualization_msgs::msg::MarkerArray & marker_array, std_msgs::msg::Header header,
    geometry_msgs::msg::Point box_min, geometry_msgs::msg::Point box_max,
    std::string obj_name, int & marker_id)
  {
    auto name_id_text_marker =
      createNameIDMarker(header, box_min, box_max, obj_name, ++marker_id);
    auto min_text_marker = createTextMarker(header, box_min, "Min", ++marker_id);
    auto max_text_marker = createTextMarker(header, box_max, "Max", ++marker_id);
    auto box_line_marker = createBoxLineMarker(header, box_min, box_max, ++marker_id);

    marker_array.markers.emplace_back(min_text_marker);
    marker_array.markers.emplace_back(max_text_marker);
    marker_array.markers.emplace_back(name_id_text_marker);
    marker_array.markers.emplace_back(box_line_marker);
    RCLCPP_DEBUG(this->get_logger(),
      "Marker: name=%s, min(%.2f,%.2f,%.2f),max(%.2f,%.2f,%.2f)",
      obj_name.c_str(), box_min.x, box_min.y, box_min.z, box_max.x, box_max.y, box_max.z);
  }

  /* Name and ID marker */
  visualization_msgs::msg::Marker createNameIDMarker(
    std_msgs::msg::Header header,
    geometry_msgs::msg::Point box_min,
    geometry_msgs::msg::Point box_max,
    std::string & name, int & marker_id)
  {
    auto marker = visualization_msgs::msg::Marker();
    marker.header = header;
    marker.id = marker_id;

    marker.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
    marker.action = visualization_msgs::msg::Marker::ADD;

    marker.scale.z = 0.05;

    marker.color.a = 1.0;
    marker.color.r = 0.0;
    marker.color.g = 1.0;
    marker.color.b = 0.0;
    std::string name_text = name;
    marker.text = name_text;

    marker.pose.position.x = (box_min.x + box_max.x) / 2;
    marker.pose.position.y = (box_min.y + box_max.y) / 2;
    marker.pose.position.z = (box_min.z + box_max.z) / 2;
    return marker;
  }

  /* Min and Max Text Marker */
  visualization_msgs::msg::Marker createTextMarker(
    std_msgs::msg::Header header,
    geometry_msgs::msg::Point position,
    const std::string & name, int & marker_id)
  {
    auto marker = visualization_msgs::msg::Marker();
    marker.header = header;
    marker.id = marker_id;

    marker.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
    marker.action = visualization_msgs::msg::Marker::ADD;

    marker.scale.z = 0.05;

    marker.color.a = 1.0;
    marker.color.r = 0.0;
    marker.color.g = 1.0;
    marker.color.b = 0.0;

    char pos[3][20];
    snprintf(pos[0], sizeof(pos[0]), "%.2f", position.x);
    snprintf(pos[1], sizeof(pos[1]), "%.2f", position.y);
    snprintf(pos[2], sizeof(pos[2]), "%.2f", position.z);

    std::string pos_x = pos[0];
    std::string pos_y = pos[1];
    std::string pos_z = pos[2];
    std::string _text = name + "[" + pos_x + "," + pos_y + "," + pos_z + "]";
    marker.text = _text;

    marker.pose.position.x = position.x;
    marker.pose.position.y = position.y;
    marker.pose.position.z = position.z;

    return marker;
  }

  /* Box Line Marker */
  visualization_msgs::msg::Marker createBoxLineMarker(
    std_msgs::msg::Header header,
    geometry_msgs::msg::Point position_min,
    geometry_msgs::msg::Point position_max,
    int & marker_id)
  {
    auto marker = visualization_msgs::msg::Marker();
    marker.header = header;
    marker.ns = "marker_test_line_list";
    marker.id = marker_id;

    marker.type = visualization_msgs::msg::Marker::LINE_LIST;
    marker.action = visualization_msgs::msg::Marker::ADD;

    marker.scale.x = 0.005;

    marker.color.r = 1.0;
    marker.color.g = 1.0;
    marker.color.b = 0.0;
    marker.color.a = 1.0;

    marker.pose.orientation.x = 0.0;
    marker.pose.orientation.y = 0.0;
    marker.pose.orientation.z = 0.0;
    marker.pose.orientation.w = 1.0;

    geometry_msgs::msg::Point p1 = position_min;

    geometry_msgs::msg::Point p2 = p1;
    p2.x = position_max.x;

    geometry_msgs::msg::Point p3 = p2;
    p3.z = position_max.z;

    geometry_msgs::msg::Point p4 = p3;
    p4.x = position_min.x;

    geometry_msgs::msg::Point p5 = p1;
    p5.y = position_max.y;

    geometry_msgs::msg::Point p6 = p2;
    p6.y = position_max.y;

    geometry_msgs::msg::Point p7 = p3;
    p7.y = position_max.y;

    geometry_msgs::msg::Point p8 = p4;
    p8.y = position_max.y;

    // draw lines p1 -> p2 -> p3 -> p4
    marker.points.push_back(p1);
    marker.points.push_back(p2);
    marker.points.push_back(p2);
    marker.points.push_back(p3);
    marker.points.push_back(p3);
    marker.points.push_back(p4);
    marker.points.push_back(p4);
    marker.points.push_back(p1);

    // draw lines p5 -> p6 -> p7 -> p8
    marker.points.push_back(p5);
    marker.points.push_back(p6);
    marker.points.push_back(p6);
    marker.points.push_back(p7);
    marker.points.push_back(p7);
    marker.points.push_back(p8);
    marker.points.push_back(p8);
    marker.points.push_back(p5);

    // draw lines 1:5, 2:6, 3:7, 4:8
    marker.points.push_back(p1);
    marker.points.push_back(p5);
    marker.points.push_back(p2);
    marker.points.push_back(p6);
    marker.points.push_back(p3);
    marker.points.push_back(p7);
    marker.points.push_back(p4);
    marker.points.push_back(p8);

    return marker;
  }

  /* localization callback for performance test */
  void loc_performance_callback(const LocalizationMsg::SharedPtr msg)
  {
    struct timespec time_start = {0, 0};
    clock_gettime(CLOCK_REALTIME, &time_start);
    static double last_sec = 0;
    static double last_nsec = 0;
    static int count = 0;
    double interval = 0;
    double current_sec = time_start.tv_sec;
    double current_nsec = time_start.tv_nsec;
    double msg_sec = msg->header.stamp.sec;
    double msg_nsec = msg->header.stamp.nanosec;

    count++;
    interval = (current_sec - last_sec) + ((current_nsec - last_nsec) / 1000000000);
    if (last_sec == 0) {
      last_sec = current_sec;
      last_nsec = current_nsec;
      return;
    }

    if (interval >= 1.0) {
      double latency = (current_sec - msg_sec) + ((current_nsec - msg_nsec) / 1000000000);
      double fps = count / interval;
      count = 0;
      last_sec = current_sec;
      last_nsec = current_nsec;
      RCLCPP_DEBUG(this->get_logger(), "L: fps %.3f hz, latency %.3f sec", fps, latency);
      loc_fps_ = fps;
      loc_latency_ = latency;
    }
  }

  /* tracking callback for performance test */
  void tra_callback(const TrackingMsg::SharedPtr msg)
  {
    struct timespec time_start = {0, 0};
    clock_gettime(CLOCK_REALTIME, &time_start);
    static double last_sec = 0;
    static double last_nsec = 0;
    static int count = 0;
    double interval = 0;
    double current_sec = time_start.tv_sec;
    double current_nsec = time_start.tv_nsec;
    double msg_sec = msg->header.stamp.sec;
    double msg_nsec = msg->header.stamp.nanosec;

    count++;
    interval = (current_sec - last_sec) + ((current_nsec - last_nsec) / 1000000000);
    if (last_sec == 0) {
      last_sec = current_sec;
      last_nsec = current_nsec;
      return;
    }

    if (interval >= 1.0) {
      double latency = (current_sec - msg_sec) + ((current_nsec - msg_nsec) / 1000000000);
      double fps = count / interval;
      count = 0;
      last_sec = current_sec;
      last_nsec = current_nsec;
      RCLCPP_DEBUG(this->get_logger(), "T: fps %.3f hz, latency %.3f sec", fps, latency);
      tra_fps_ = fps;
      tra_latency_ = latency;
    }
  }
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MarkerPublisher>());
  rclcpp::shutdown();
  return 0;
}
